### PR TITLE
Consider nextIntervalRef of other half of double registers for ARM

### LINKED ISSUE #46086: Assertion failed '!"Spill candidate has no assignedInterval recentRefPosition"' in 'System.Numerics.Tests.Vector2Tests:Vector2LerpTest7():this' during 'LSRA build intervals'
https://dev.azure.com/dnceng/public/_build/results?buildId=921374&view=ms.vss-test-web.build-test-results-tab&runId=29267922&paneView=debug&resultId=173687

net6.0-Linux-Release-arm-CoreCLR_checked-jitstress2_jitstressregs0x80
net6.0-Linux-Release-arm-CoreCLR_checked-jitstress2_jitstressregs3

System.Numerics.Vectors.Tests

```
Assert failure(PID 26 [0x0000001a], Thread: 41 [0x0029]): Assertion failed '!"Spill candidate has no assignedInterval recentRefPosition"' in 'System.Numerics.Tests.Vector2Tests:Vector2LerpTest7():this' during 'LSRA build intervals' (IL size 71)

    File: /__w/1/s/src/coreclr/jit/lsra.cpp Line: 3496
    Image: /root/helix/work/correlation/dotnet
```

category:correctness
theme:register-allocator
skill-level:expert
cost:small

### PR DESCRIPTION
**Pull request by kunalspathak**
In the recent https://github.com/dotnet/runtime/pull/45135, we changed the way we go through the heuristics of selecting registers. If we decide to use a busy register, using `FAR_NEXT_REF` heuristics, we try to pick a register whose next interval reference is farthest. While calculating the location, we were not taking into consideration the next location of other half of the pair of registers, if the interval type is `TYP_DOUBLE`. Because of this, we were adding a register in "potential candidate to spill" that doesn't even have assigned interval. Further, in `REG_NUM`  heuristics, we ended up picking this wrong register and hitting assert.

The fix was to take into account the next interval ref of other half of double register so we add only the legitimate candidates to be spilled.

Once that was fixed, there was also a problem I discovered in `PREV_REG_OPT` heuristics. In this, out of remaining candidates, we check if they have valid assigned interval and if not, we assert because they can't be spill candidates. However, if the current interval is `TYP_DOUBLE`, we should have also check if the other register of the double register pair has valid interval and if yes, just check the heuristics against its refPosition.

Fixes: https://github.com/dotnet/runtime/issues/46086

### CODE REVIEW

**Comment by kunalspathak** [comment]
arm failures are related to https://github.com/dotnet/runtime/issues/46085 and are fixed in https://github.com/dotnet/runtime/pull/46939 .

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -4590,27 +4542,42 @@ RegRecord* LinearScan::getSecondHalfRegRec(RegRecord* regRec)
...
+// Arguments:
+//    regNumber - A float regNumber
+//
+// Assumptions:
+//    None
+//
+// Return Value:
+//    A RegRecord which forms same double register with regRec
+//
+regNumber LinearScan::findAnotherHalfRegRec(regNumber regNum)
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:4561]: This function should be renamed `findDoubleRegPair` or similar: it shouldn't imply that it's finding a RegRecord.

File: src/coreclr/jit/lsra.cpp [hunk 2]
```diff
@@ -3479,22 +3388,65 @@ regNumber LinearScan::allocateReg(Interval* currentInterval, RefPosition* refPos
             prevRegOptCandidates &= ~prevRegOptCandidateBit;
             regNumber prevRegOptCandidateRegNum = genRegNumFromMask(prevRegOptCandidateBit);
             Interval* assignedInterval          = physRegs[prevRegOptCandidateRegNum].assignedInterval;
-            // The assigned should be non-null, and should have a recentRefPosition, however since
-            // this is a heuristic, we don't want a fatal error, so we just assert (not noway_assert).
+            bool      foundPrevRegOptReg        = true;
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:3391]: It looks to me like `foundPrevRegOptReg` will always be `true`; how can it be set to `false`?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/lsra.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:4561]: Sure. I have changed it to `findAnotherHalfRegNum()`

File: src/coreclr/jit/lsra.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:3391]: Yes, I went back and forth and seems I missed the case when it should be `false`. I have updated the code.

For reference, I was trying to add back the missing code that was present pre #45135,. You can see it at https://github.com/dotnet/runtime/pull/45135/files#diff-a985c84f41c0c758b9a2f849758cdfd6eac4acf6cfc2151a28d206d70e0358e0L3816-L3824 (need to load diffs for `lsra.cpp` changes).

File: src/coreclr/jit/lsra.cpp [hunk 3]
```diff
@@ -4591,26 +4549,41 @@ RegRecord* LinearScan::getSecondHalfRegRec(RegRecord* regRec)
...
+// findAnotherHalfRegNum: Find another half register's number which forms same ARM32 double register
+//
+// Arguments:
+//    regNumber - A float regNumber
+//
+// Assumptions:
+//    None
+//
+// Return Value:
+//    A register number which forms same double register with regRec
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:4565]: nit: `regRec` is wrong here.

**Review by BruceForstall — APPROVED** [review]
LGTM (if the tests pass)

File: src/coreclr/jit/lsra.cpp (hunk 3 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp]: fixed.

**Pull request merged**